### ISSUE TITLE
How does TurboPFor work?

### ISSUE THREAD

**Issue by pps83** [powturbo/TurboPFor#37]
Is TurboPFor a library for int compression? If yes, why does the page look like some compare tool for different libs. Put the docs/hello world right there, not the compare results. It's extremely painful to find anything how to use the lib (I hope it's wasn't assumed that users should read icbench or source of any other tool to decifir how to use the lib).

I use visual studio, and VS build doesn't work at all. Also, that approach that some files result in multiple different output obj files doesn't help overall: it would be better if the lib had tiny wrappers that included those c-files and set required defines to modify compilation.

Even after I made it build on VS it absolutely doesn't work and completely fails for me.
I'm not even sure what function I need to use to compress. After that huge table of irrelevant info, if anybody ever gets as far as ["Function usage"](https://github.com/powturbo/TurboPFor#function-usage) section, you'll see some cryptic explanation that doesn't seem to be current anymore. From that section, it seems that I need p4fmencXXX but that doesn't exist (nothing p4fm exist at all). Why is that there that pack/unpack thing doing, to confuse people? From the docs, it seems like p4packXXX should be a valid function, while it's not.

After digging, it seem that I should try `p4enc32`, but so far I get stack corruption, and it doesn't seem like it could ever work at all. It crashes for me in `_p4enc32` on line `while(i != n) MISS;`
In my case I call `p4enc32` with an array of 2731 uints. That while loop will clearly corrupt stack, because the `MISS` expands to `_in[i] = in[i]`, that is, it will attempt to assign `_in[2730]` because that loop will run up to the length of the input `n=2731`. I don't get it, how come that code could even work, as it writes to stack array of 287 elements?! Also, it seems like there are no checks, I don't see a single assert anywhere to check/show conditions/expectations, while in that function it had to make that check!

**Comment by powturbo** [comment]
Thanks for your detailed comments. I suggest reading [The First Rule of Programming: It's Always Your Fault](https://blog.codinghorror.com/the-first-rule-of-programming-its-always-your-fault/)
> Does TurboPFor work? 

Download the executable icapp and convince yourself, that TurboPFor not only works, but can handle nearly every existing integer input data format (raw, text, csv,...) without any conversion and without crashing.

TurboPFor is the only integer compression library than can process all FULL range integer and floating point formats (8,16,32,64 integers + float and double) including diverse integrated functions like delta,zigzag,... with or without SIMD. All that in one single package. 

Until today, no other integer compression library can compress or decompress faster with better compression. You can verify the claims, by using the benchmarking tool "icbench" including the best available external libraries.

> ... why does the page look like some compare tool for different libs

You have several sections in the main page, benchmarks, compiling, testing, usage,... 
There are over 500 functions in TurboPFor and for all the compression functions the usage is as simple as memcpy. No containers, no context pointers, no states, no initialization, no factories.
All you need is your input,output buffers + the corresponding lengths.

As stated in the usage section it is recommanded to look at the API header files "vp4.h, vint.h,vsimple.h,..." for the list and a limited documentation. 
Some low level function are limited to 128/256 integers. I suggest also to look at the file "icapp.c" or the short example in [Hello TurboPFor](https://github.com/powturbo/TurboPFor/issues/26)

> I use visual studio, and VS build doesn't work at all.

Actually there is no project file for visual studio, but there is a separate NMAKE build  file "makefile.vs" for all recent (>=VS2010) microsoft compilers.
You can compile and build a dll or a lib file. You have not to include the c files into your visual studo project.
It is also possible to use the mingw64 compiler under windows for building the libs.

> After digging, it seem that I should try p4enc32

p4enc is a low level function and is limited to 128/256 (for SSE/AVX2) integers (see the comments in vp4.h).
You should use the p4nenc function for large arrays. 
You must also allocate you buffers large enough (see [Hello TurboPFor](https://github.com/powturbo/TurboPFor/issues/26)).
Do not hesitate to open a github issue or contact me, in case of other difficulties.

**Comment by pps83** [comment]
It’s extremely painful to use a library that does not have proper function definitions. When it abuses C-macro for function definitions code becomes unsearchable. When I tried to use p4nenc, after grepping it seems that these are unoptimized versions.
In contrast, I also use FastPFor, I got it all working with no issues, results are amazing and I simply wanted to compare with TurboPFor and I cannot even see what should I use.

**Comment by pps83** [comment]
I don’t even see what uint_t stands for. What a mess with macro defines

**Comment by pps83** [comment]
Just in case it’s not obvious for you, I want to point out that it does not make sense that a library that does number compression includes windows.h from its public header. Impl may use it, but not a header!

Including a header from TurboPFor should not pollute my namespace with all those macro madness that has nothing to do with number compression. All these AS AC die sleep etc shouldn’t be exposed to those who use it. I assume those macro were meant for internal use of the lib and not for the users of the lib. If I include turbopfor headers I get tons of errors because of all these defines. Also, it doesn’t even compile by itself. Just try to compile empty file that includes vp4.h (on windows, not sure if it’s the same broken elsewhere) and it will error out on undefined math functions.

In short, a properly designed library should not expose its messy guts to the users of the lib.

**Comment by pps83** [comment]
After I undefined all those macro that broke compilation at the end Compiler tells me that p4nenc and p4ndec are not defined.

**Comment by pps83** [comment]
I have no idea what to do. You say I need to use p4nenc, which does not exists. Yet, helloworld that you are referring to uses that function that I tried to use originally.

I simply want to compare with FastPFro. I have an array of uint32_t values and I want to compress it and then uncompress it. That’s all. There is nothing easier than that, yet, it seems to be impossible to get basics running

**Comment by pps83** [comment]
If I assume that you unintentionally made a mistake suggesting to use p4nenc/p4ndec, and instead I had to use p4nenc32/p4ndec32, then I get the following results: 1) decoder doesn't produce identical result (to original input). 2) it crashes for me in the middle of unreadable macrofest with `Access violation executing location 0x0000000000000000.`:


![image](https://user-images.githubusercontent.com/1614246/61207655-8525aa00-a6aa-11e9-8e6c-d35de1ba351a.png)

**Comment by powturbo** [comment]
As already mentioned you have only to use the public header file "vp4.h" "bitpack.h" for the TurboPFor  or bitpack functions. You have not to look at the source code. 
With "p4nenc", I'm refering to the scalar "p4nenc8,p4nenc16,p4nenc32,p4nenc64" and the vectorized "p4nenc128v16, p4nenc128v32, p4nenc128v64" or "p4nenc256v32" (see file "vp4.h") depending on the integer size 8,16,32 or 64 bits.  In your case it is "p4nenc32" or "p4nenc128v32" or "p4nenc256v32".

For comparison purpose, you can also download the executable "icapp" for windows and test TurboPFor with your own data.
In case of crashing, check the parameter sequence and the buffer sizes. The input/output buffer size should be slightly larger  (see [Hello TurboPFor](https://github.com/powturbo/TurboPFor/issues/26)).

**Comment by pps83** [comment]
In my case I'm looking for integer compressor that has compresses well and decompresses fast (eg I compress once, and then decompress many times). I have a test that tries all available codecs to compare results and even if I ignore the fact that p4enc32 doesn't produce correct result and exclude those inputs that crash it, then I get these results:
```
   0:615 === results
   0:615 copy                    :  0.00%,  105 nanosec. 10924 =>  10924 bytes. Decompress:( 105  315  154)
   0:615 simdbinarypacking       : 73.34%,  373 nanosec. 10924 =>   2911 bytes. Decompress:( 373  962  460)
   0:616 streamvbyte             : 64.33%,  519 nanosec. 10924 =>   3896 bytes. Decompress:( 519 1141  630)
   0:616 varintg8iu              : 66.33%,  525 nanosec. 10924 =>   3677 bytes. Decompress:( 525 1074  612)
   0:616 simdnewpfor             : 74.84%,  573 nanosec. 10924 =>   2748 bytes. Decompress:( 573 1670  707)
   0:616 simdsimplepfor          : 76.22%,  628 nanosec. 10924 =>   2597 bytes. Decompress:( 628 7866 1181)
   0:617 simdpfor                : 57.10%,  703 nanosec. 10924 =>   4686 bytes. Decompress:( 703 1331  791)
   0:617 simdgroupsimple_ringbuf : 70.43%,  916 nanosec. 10924 =>   3230 bytes. Decompress:( 916 1885 1050)
   0:617 simdgroupsimple         : 72.79%,  922 nanosec. 10924 =>   2972 bytes. Decompress:( 922 1722 1041)
   0:617 maskedvbyte             : 67.91%,  989 nanosec. 10924 =>   3505 bytes. Decompress:( 989 2154 1134)
   0:618 simdfastpfor128         : 75.98%, 1179 nanosec. 10924 =>   2624 bytes. Decompress:(1179 2640 1355)
   0:618 varintgb                : 64.33%, 1225 nanosec. 10924 =>   3896 bytes. Decompress:(1225 1974 1353)
   0:618 simdoptpfor             : 76.66%, 1335 nanosec. 10924 =>   2549 bytes. Decompress:(1335 2465 1485)
   0:618 p4                      : 77.24%, 1361 nanosec. 10924 =>   2486 bytes. Decompress:(1361 4629 1665)
   0:618 simdfastpfor256         : 74.77%, 1427 nanosec. 10924 =>   2756 bytes. Decompress:(1427 2868 1617)
   0:619 fastbinarypacking16     : 74.39%, 1506 nanosec. 10924 =>   2797 bytes. Decompress:(1506 3056 1695)
   0:619 BP32                    : 75.55%, 1687 nanosec. 10924 =>   2670 bytes. Decompress:(1687 3265 1866)
   0:619 fastbinarypacking32     : 75.55%, 1695 nanosec. 10924 =>   2670 bytes. Decompress:(1695 2968 1855)
   0:619 simplepfor              : 76.22%, 1844 nanosec. 10924 =>   2597 bytes. Decompress:(1844 9462 2456)
   0:620 newpfor                 : 74.84%, 1922 nanosec. 10924 =>   2748 bytes. Decompress:(1922 3574 2134)
   0:620 fastbinarypacking8      : 70.59%, 1991 nanosec. 10924 =>   3213 bytes. Decompress:(1991 4198 2276)
```

the table is sorted by fastest decompression speed (memcpy being fastest). p4 compresses marginally better than some of the best FastPFor codecs, but decompresses slower `Decompress` output displays (fastest decompression, slowest, average).

It seems that some codecs consistently hit slow path on some types of inputs when decoding. `simdsimplepfor` has this problem: its fastest decode time is 628ns, while average is almost twice as large:1181ns. Even though it's the closes to p4 by compression ratio, overall I get much better results with `simdfastpfor128` because of its consistent decoding speed.

The test goes through input arrays that I need to compress and tries to decompress them 10 times in a loop and computes fastest/slowest/average time.

**Comment by pps83** [comment]
> For comparison purpose, you can also download the executable "icapp" for windows and test TurboPFor with your own data.
> In case of crashing, check the parameter sequence and the buffer sizes. The input/output buffer size should be slightly larger (see [Hello TurboPFor](https://github.com/powturbo/TurboPFor/issues/26)).

The data that goes through test is dynamically generated. In my case it's always array of 2731 32-bit unsigned ints. For the buffer of compressed data I use 2731*2 of uint32_t, same buffer size I use for decompression (that is, it's way more that needed).
If I use p4nenc128v32/p4ndec128v32 I get the same crash in p4ndec128v32 on some of the input. If I exclude data that crashes it's approximately 15-20% faster according to my tests (I measure decoder speed only).

**Comment by pps83** [comment]
`p4nenc256v32` fails for me on `kmovw` opcode (I have latest macbook pro, not sure what cpu it has). If I disable `__AVX512F__` switch, then this problem goes away. However, `p4ndec256v32` fails with that same error mentioned previously on input that don't trigger that crash with p4ndec32/p4ndec128v32

**Comment by pps83** [comment]
This might possibly point out why it fails:

![image](https://user-images.githubusercontent.com/1614246/61212693-dee0a100-a6b7-11e9-99ca-b1f666594fc3.png)

As you can see `v32compSize` was 2805 bytes. `p4ndec128v32` returns 2880. For some reason it reads 75 bytes more. I get identical numbers with non-simd versions. The way I manage buffers all those trailing data that the functions read should be filled with zeros.

This is the input array that produces wrong results after it's encoded and then decoded:
https://gist.github.com/pps83/6f1a6bcd6c4df464a3d90a69c24592b5

After the compressor/decompressor runs I get difference in the last byte only.

**Comment by powturbo** [comment]
TurboPFor can read/write beyond buffer end. You must reserve/allocate memory (32 items) larger than necessary (again, please see the Hello TurboPFor example).

**Comment by pps83** [comment]
> TurboPFor can read/write beyound buffer end. You must reserve/allocate memory (32 items) larger than necessary (again, please see the Hello TurboPFor example).

As I've already mentioned, to compress 2731 uint32_t items, I allocate 5462 uint32_t items for the compressed output, and I allocated 5462 of uint32_t items for decompressed output. Like 100 times more than p4 lib expects.

**Comment by powturbo** [comment]
I've done a quick look at your coding. You should always use the number of the array elements for encoding or decoding. p4ndec128v32 should be called with dataSize instead of v32compSize. The Input buffer must be also (32 items) larger. Instead of "const unsigned testData[] =", try "const unsigned testData[2731+32] ="

**Comment by pps83** [comment]
I don't use testData, I just printed it for reference. `dataSize` from that screenshot is the number of uint32_t elements.
ah, ok, I see what you are saying, decoder should take not input size in bytes, but the output size in uint32_t?.. that would explain why I get all these issues. I'll retry and report back.
Thanks

**Comment by pps83** [comment]
Quick update: with suggested fix to pass decoded size instead of encoded I got it all working, no mroe crashes or corruptions (I had to disable `__AVX512F__` for 256v version though). More over, compression rate is noticeably better than what I got with FastPFor. I got on average 83.04% compression rate with 256v version, and 83.60% with 128v and regular version version. 128v version however was noticeably by 40% slower than simdfastpfor128 from FPF (while regular version is only 10% slower). 256v version compresses a bit worse than 128v and regular versions, but has the same decompression speed as `simdfastpfor128`. `simdfastpfor128` compression rate is 81.45%, which might not seem like a big difference; however, 81.45% vs 83.04% rate results in compression sizes of 2026 vs 1853 bytes, which is almost a 10% difference.

Side note: it's quite unusual that the decompressor expects decompressed size as the input instead of compressed buffer size (other codecs that I've used, all FastPFor codecs, lz4 etc, all expect compressed size).

**Comment by powturbo** [comment]
TurboPFor can efficiently compress blocks as small as 128/256 integers in size (the scalar version even smaller). Other integer compressors like FastPFor or general purpose compressors are inefficient and slow for small blocks. 
Small blocks are important in query processors like those in search engines. It permits to
decompress only blocks involved in the query evaluation.
If you want to compress large arrays only, why not simply use a transpose/shuffle + lz4 for example? (see [TurboTranspose](https://github.com/powturbo/TurboTranspose) or try "icapp") .

TurboPFor is also adaptive to different data distributions. see [TurboPFor: an analysis](https://michael.stapelberg.ch/posts/2019-02-05-turbopfor-analysis/). 
TurboPFor supports FULL range 8,16,32,64 bits integers + floating point (float,double), has
several integrated codecs like delta, delta one,zigzag,for,..., binary compatible ARM NEON support, Java interface, ++++ ALL that in a SINGLE package.

> it's quite unusual that the decompressor expects

This is intentional, in this case you don't need to additionally store the compressed length (saving 4 bytes). Important for small blocks.

**Comment by pps83** [comment]
I'll try transpose/shuffle + lz4 and TurboTranspose to see what results I get. But from my testing lz4 was much worse in compression and most importantly decoder speed compared to other codecs. I even excluded it from my tests as it makes my test take too long.

The data that I compress is an array of 390 structs, where each struct contains 5 closely related ints, and the two other unrelated numbers. The sample data array (posted as gist previously) contains that data, but in already preprocessed state: the 5 ints are extracted, with first number of the entire array is the base, and all other numbers (5*390) are simply difference from that first number. Then it goes by two other arrays of 390 ints each.

**Comment by pps83** [comment]
I added some VS-specific fixes in [msvc fixes PR](https://github.com/powturbo/TurboPFor/pull/40)

**Comment by pps83** [comment]
I saved sample binary data and tried to run icapp and I get this:
```
   E MB/s     size     ratio   D MB/s  function (integer size=32 bits)
  1960.72      54918  23.94% 11470.20 p4nenc32         TurboPFor                        testdata.bin
  2104.62      54918  23.94% 22940.40 p4nenc128v32     TurboPForV                       testdata.bin
  2317.21      56407  24.59% 28675.50 p4nenc256v32     TurboPFor256                     testdata.bin
  1201.07     118770  51.77%  4170.98 p4ndenc32        TurboPFor    delta               testdata.bin
  1220.23     118770  51.77%  6200.11 p4ndenc128v32    TurboPForV   delta               testdata.bin
  1461.17     119745  52.20%  7910.48 p4ndenc256v32    TurboPFor256 delta               testdata.bin
  1102.90     157711  68.75%  3823.40 p4nd1enc32       TurboPFor    delta 1             testdata.bin
  1102.90     157711  68.75%  5462.00 p4nd1enc128v32   TurboPForV   delta 1             testdata.bin
  1274.47     158496  69.09%  6747.18 p4nd1enc256v32   TurboPFor256 delta 1             testdata.bin
  1509.24      44941  19.59%  4987.04 p4nzenc32        TurboPFor    zigzag              testdata.bin
  1560.57      44941  19.59%  7646.80 p4nzenc128v32    TurboPForV   zigzag              testdata.bin
  1792.22      46378  20.22%  9558.50 p4nzenc256v32    TurboPFor256 zigzag              testdata.bin
 11470.20      61161  26.66% 14337.75 bitnpack32       TurboPack                        testdata.bin
 20854.91      61161  26.66% 38234.00 bitnpack128v32   TurboPackV                       testdata.bin
 32772.00      67977  29.63% 45880.80 bitnpack256v32   TurboPack256                     testdata.bin
  9176.16     229851 100.19% 13494.35 bitndpack32      TurboPack    delta               testdata.bin
 13494.35     229851 100.19% 19117.00 bitndpack128v32  TurboPackV   delta               testdata.bin
 16386.00     229627 100.10% 19117.00 bitndpack256v32  TurboPack256 delta               testdata.bin
  8823.23     229851 100.19% 12744.67 bitnd1pack32     TurboPack    delta 1             testdata.bin
 14337.75     229851 100.19% 17646.46 bitnd1pack128v32 TurboPackV   delta 1             testdata.bin
 14337.75     229627 100.10% 16386.00 bitnd1pack256v32 TurboPack256 delta 1             testdata.bin
  4779.25      56826  24.77%  6200.11 bitnzpack32      TurboPack    zigzag              testdata.bin
  9558.50      56826  24.77% 13494.35 bitnzpack128v32  TurboPackV   zigzag              testdata.bin
 12073.89      66442  28.96% 17646.46 bitnzpack256v32  TurboPack256 zigzag              testdata.bin
  9176.16      65950  28.75% 10427.45 vbenc32          TurboVByte                       testdata.bin
  5213.73      65663  28.62%  5735.10 vbzenc32         TurboVByte   zigzag              testdata.bin
  1880.36      57130  24.90%  6372.33 vsenc32          TurboVSimple                     testdata.bin
  1461.17      44686  19.48%  3760.72 vszenc32         TurboVSimple zigzag              testdata.bin
  1310.88      69211  30.17%  1686.79 bvzzenc32        bitio zigzag of delta            testdata.bin
  1240.02      61632  26.87%  1442.79 bvzenc32         bitio zigzag                     testdata.bin
  1318.41      61342  26.74%  1451.92 fpgenc32         bitio TurboGorilla               testdata.bin
  1318.41      51007  22.23%  3142.52 fpzzenc32        TurboPFor zigzag of delta        testdata.bin
  1164.49      44947  19.59%  1499.37 fpfcmenc32       TurboPFor FCM                    testdata.bin
  1130.07      47540  20.72%  1349.44 fpdfcmenc32      TurboPFor DFCM                   testdata.bin
  1124.53      48680  21.22%  1318.41 fp2dfcmenc32     TurboPFor DFCM 2D                testdata.bin
   469.13     226485  98.73% 25489.33 trle             TurboRLE                         testdata.bin
   284.97     226948  98.93%  1686.79 trlez            TurboRLE   zigzag                testdata.bin
  1764.65     218516  95.25% 22940.40 srle32           TurboRLE32 escape                testdata.bin
  2066.70     222994  97.21% 11470.20 srlez32          TurboRLE32 escape                testdata.bin
 32772.00     229404 100.00% 28675.50 memcpy           memcpy                           testdata.bin
```

Seems like `p4nzenc32` compresses better, however, 2-3 times slower when decompressing.
How can I test TurboTranspose with icapp?

**Comment by powturbo** [comment]
> How can I test TurboTranspose with icapp?

compile with:
"make CODEC1=1 CODEC2=1"

**Comment by powturbo** [comment]
The timings are meaningless when you benchmark small length data. You will get the timings for the L1/L2 caches. In practice, it is unlikely that your data stays in the L1/L2 cache.

**Comment by pps83** [comment]
Did you review https://github.com/powturbo/TurboPFor/pull/40, any comments?

Another interesting observation: One of the best codecs from [FastPFor](https://github.com/lemire/FastPFor) `SimdFastPFor128`  adds 182KB to my binary when both encoder and decoder are used, whereas p4nenc256v32+p4ndec256v32 add roughly 1636KB, and p4nzenc256v32+p4nzdec256v32 add roughly 2107KB, which is surprisingly 10x from comparable codec from FastPFor lib.

**Comment by powturbo** [comment]
> Did you review #40, any comments?

Thanks for your adaptation. I have no access to my home pc until next week.

> Another interesting observation...

I'll look a this next time, but TurboPFor is using extensive unrolling. The AVX2 functions are
also including the "scalar TurboPFor" + "scalar bitpacking" + "variable byte" functionsto process variable length arrays (not multiple of 256). In FastPFor you must use the composite codecs.
Additionally huge static  AVX2 permutation tables are also included.

p4nzenc256v32+p4nzdec256v32 are integrated codecs. In this case (to save binary space), it is also possible to use the bitzenc/bitzdec from "bitutil.h" in conjunction with p4nenc256v32+p4ndec256v32.
As already stated, TurboPFor is adaptive to the distribution of the data and is dynamically switching between several base codecs.

**Comment by powturbo** [comment]
Pull request commited. Thanks for your valuable contribution.
